examples/inspection/plot_permutation_importance.py

```diff
@@ -121,59 +130,109 @@
-    result.importances[sorted_idx].T, vert=False, labels=X_test.columns[sorted_idx]
+sorted_importances_idx = result.importances_mean.argsort()
+importances = pd.DataFrame(
+    result.importances[sorted_importances_idx].T,
+    columns=X.columns[sorted_importances_idx],
 )
-ax.set_title("Permutation Importances (test set)")
-fig.tight_layout()
-plt.show()
+importances.plot.box(vert=False, whis=10)
+plt.title("Permutation Importances (test set)")
+plt.axvline(x=0, color="k", linestyle="--")
+plt.xlabel("Decrease in accuracy score")
+plt.tight_layout()
 
 # %%
 # It is also possible to compute the permutation importances on the training
-# set. This reveals that ``random_num`` gets a significantly higher importance
-# ranking than when computed on the test set. The difference between those two
-# plots is a confirmation that the RF model has enough capacity to use that
-# random numerical feature to overfit. You can further confirm this by
-# re-running this example with constrained RF with min_samples_leaf=10.
+# set. This reveals that `random_num` and `random_cat` get a significantly
+# higher importance ranking than when computed on the test set. The difference
+# between those two plots is a confirmation that the RF model has enough
+# capacity to use that random numerical and categorical features to overfit.
 result = permutation_importance(
     rf, X_train, y_train, n_repeats=10, random_state=42, n_jobs=2
 )
-sorted_idx = result.importances_mean.argsort()
 
-fig, ax = plt.subplots()
-ax.boxplot(
-    result.importances[sorted_idx].T, vert=False, labels=X_train.columns[sorted_idx]
+sorted_importances_idx = result.importances_mean.argsort()
+importances = pd.DataFrame(
+    result.importances[sorted_importances_idx].T,
+    columns=X.columns[sorted_importances_idx],
 )
-ax.set_title("Permutation Importances (train set)")
-fig.tight_layout()
-plt.show()
+importances.plot.box(vert=False, whis=10)
+plt.title("Permutation Importances (train set)")
+plt.axvline(x=0, color="k", linestyle="--")
+plt.xlabel("Decrease in accuracy score")
+plt.tight_layout()
+
+# %%
+# We can further retry the experiment by limiting the capacity of the trees
+# to overfit by setting `min_samples_leaf` at 20 data points.
+rf.set_params(classifier__min_samples_leaf=20).fit(X_train, y_train)
+
+# %%
+# Observing the accuracy score on the training and testing set, we observe that
+# the two metrics are very similar now. Therefore, our model is not overfitting
+# anymore. We can then check the permutation importances with this new model.
+print(f"RF train accuracy: {rf.score(X_train, y_train):.3f}")
+print(f"RF test accuracy: {rf.score(X_test, y_test):.3f}")
+
+# %%
+importances = {}
+for name, data, target in zip(["train", "test"], [X_train, X_test], [y_train, y_test]):
+    result = permutation_importance(
+        rf, data, target, n_repeats=10, random_state=42, n_jobs=2
+    )
+    if name == "train":
+        sorted_importances_idx = result.importances_mean.argsort()
```

Comment:
I see that we want `test` to have the same order as `train`. With that in mind, I think it's easier to follow to be verbose here:

```python
train_result = permutation_importance(
    rf, X_train, y_train, n_repeats=10, random_state=42, n_jobs=2
)
test_results = permutation_importance(
    rf, X_test, y_test, n_repeats=10, random_state=42, n_jobs=2
)
sorted_importances_idx = train_result.importances_mean.argsort()

train_importances = pd.DataFrame(
    train_result.importances[sorted_importances_idx].T,
    columns=X.columns[sorted_importances_idx],
)
test_importances = pd.DataFrame(
    test_results.importances[sorted_importances_idx].T,
    columns=X.columns[sorted_importances_idx],
)

# plotting
for name, importances in zip(["train", "test"], [train_importances, test_importances]):
    ax = importances.plot.box(vert=False, whis=10)
```

### Alternative

If you want the for loops:

```python
results = {}
for name, data, target in zip(["train", "test"], [X_train, X_test], [y_train, y_test]):
    results[name] = permutation_importance(
        rf, data, target, n_repeats=10, random_state=42, n_jobs=2
    )

importances = {}
sorted_importances_idx = results["train"].importances_mean.argsort()
for name, result in results.items():
    importances[name] = pd.DataFrame(
        result.importances[sorted_importances_idx].T,
        columns=X.columns[sorted_importances_idx],
    )

for name in ["train", "test"]:
    ax = importances[name].plot.box(vert=False, whis=10)
```

Basically, trying to avoid the `reset_index` and `groupby` if possible.